Tactics/Reflect/AxEffects.lean

```diff
@@ -0,0 +1,590 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Alex Keizer
+-/
+
+import Arm.State
+import Tactics.Common
+import Tactics.Attr
+import Std.Data.HashMap
+
+open Lean Meta
+
+/-- A reflected `ArmState` field, see `AxEffects.fields` for more context -/
+structure AxEffects.FieldEffect where
+  value : Expr
+  /-- A proof that `r <field> <currentState> = <value>` -/
+  proof : Expr
+  deriving Repr
+open AxEffects.FieldEffect
+
+/-- `AxEffects` is an axiomatic representation of effects,
+i.e., `AxEffects` transforms a description of an `ArmState` written as
+a sequence of `w` and `write_mem`s to some initial state
+into a set of hypotheses that relates reading fields from the final state
+to the initial state.
```

Comment:
I've added the following comment both here and at `fromExpr`:
```lean
Note that as soon as an unsupported expression (e.g., an `if`) is encountered,
the whole expression is taken to be the initial state,
even if there might be more `w`/`write_mem`s in sub-expressions.
```

However, `sym_n`'s implementation does additionally check that the initial state is what we expect, so added a comment there as well:
```
/- Assert that the initial state of the obtained `AxEffects` is equal to
    the state tracked by `c`.
    This will catch and throw an error if the semantics of the current
    instruction still contains unsupported constructs (e.g., an `if`) -/
```